components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +87,166 @@ public AbstractHTTPFunction() {
+                            .resultMessage("Successfully called the external endpoint.")
+                            .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                            .resultStatus(DiagnosticLog.ResultStatus.SUCCESS);
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
                 }
+                return result;
+            }
+        }
+        return Pair.of(outcome, json); // Return outcome and json (which might be null if never successful)
+    }
 
-            } catch (IllegalArgumentException e) {
+    /**
+     * Execute the request.
+     *
+     * @param request     HttpUriRequest.
+     * @param endpointURL Endpoint URL.
+     * @return Pair of outcome and json.
+     */
+    private Pair<String, JSONObject> executeRequest(HttpUriRequest request, String endpointURL) {
+
+        JSONObject json = null;
+        String outcome;
+
+        try (CloseableHttpResponse response = client.execute(request)) {
+            int responseCode = response.getStatusLine().getStatusCode();
+            if (responseCode >= 200 && responseCode < 300) {
+                outcome = Constants.OUTCOME_SUCCESS;
+                if (response.getEntity() != null) {
+                    Header contentType = response.getEntity().getContentType();
+                    String jsonString = EntityUtils.toString(response.getEntity());
+                    if (contentType != null && contentType.getValue().contains(TYPE_TEXT_PLAIN)) {
+                        json = new JSONObject();
+                        json.put(RESPONSE, jsonString);
+                    } else {
+                        JSONParser parser = new JSONParser();
+                        json = (JSONObject) parser.parse(jsonString);
+                    }
+                }
+                return Pair.of(outcome, json); // Success, return immediately
+            } else if (responseCode == HTTP_STATUS_UNAUTHORIZED) {
+                LOG.warn("Received 401 response from external API call. Url: " + endpointURL);
+                return Pair.of(Constants.OUTCOME_FAIL, null); // Unauthorized, no retry
+            } else {
+                outcome = Constants.OUTCOME_FAIL; // Other client or server error, retry if attempts left
+            }
+        } catch (Exception e) {
+            // Log the error based on its type
+            if (e instanceof IllegalArgumentException) {
                 LOG.error("Invalid Url: " + endpointURL, e);
-                outcome = Constants.OUTCOME_FAIL;
-            } catch (ConnectTimeoutException e) {
+            } else if (e instanceof ConnectTimeoutException) {
                 LOG.error("Error while waiting to connect to " + endpointURL, e);
-                outcome = Constants.OUTCOME_TIMEOUT;
-            } catch (SocketTimeoutException e) {
+            } else if (e instanceof SocketTimeoutException) {
                 LOG.error("Error while waiting for data from " + endpointURL, e);
```

Comment:
Utilize timeout related exceptions to decide whether to retry and also to improve the context of the log

Reply:
Fixed with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/05c4f77e8d18ba911c3b98bc7e18ec3fb73dcd33